### ISSUE TITLE
no-use-before-declare: error if function is used before declaration

### ISSUE THREAD

**Issue by OliverJAsh** [palantir/tslint#1537]
When using `no-use-before-declare`, I expected an error given

```
foo(); // error!
function foo() {}
```

`foo` is hoisted and so this code works, but it can lead to runtime errors, e.g.

```
foo(); // RangeError: baz is not defined
const baz = [1]
function foo() { baz; }
```

I think ESLint has an option for this

**Comment by adidahiya** [comment]
ok, we'd accept a PR for a rule option that is lenient for function declarations called something like `"allow-functions"`. this seems low priority since this is one of the less useful rules in TSLint core.

**Comment by OliverJAsh** [comment]
I would like to move to using hoisted function declarations, however currently there is no way to guarantee safety with tslint, unlike with ESLint:

``` js
// ESLint
{
    // Function called before definition and assignment
    // Static error: TS: used before assignment
    // Static error: ESLint: no-usage-before-define
    foo()
    const x = 1;
    const foo = () => { x; }
}

{
    // Function called before declaration
    // Static error: ESLint: no-usage-before-define
    foo()
    const x = 1;
    function foo() { x; }
}

{
    // Function dependencies defined after function
    // Static error: ESLint: no-usage-before-define
    function foo() { x; }
    const x = 1;
    foo()
}
```

Compare this to tslint:

``` js
// TSLint
{
    // Function called before definition and assignment
    // Static error: TS: used before assignment
    // Static error: ESLint: no-usage-before-define
    foo()
    const x = 1;
    const foo = () => { x; }
}

{
    // Function called before declaration
    // No static error
    foo2()
    const x = 1;
    function foo2() { x; }
}

{
    // Function dependencies defined after function
    // No static error
    function foo3() { x; }
    const x = 1;
    foo3()
}
```

I think we should make `no-use-before-declare` behave just like ESLint's well defined `no-use-before-define`. What do you think?

**Comment by adidahiya** [comment]
> I think we should make no-use-before-declare behave just like ESLint's well defined no-use-before-define.

Sure, I'd say go for it... but I still hold that this is a low priority core rule because compiler checks and CFA do a good job of checking for the code patterns you're looking to lint here, essentially obviating the rule. We don't enable it in the `tslint:recommended` configuration.

**Comment by OliverJAsh** [comment]
In the bottom example above, the last two examples are not caught by TS. I opened an issue about that at https://github.com/Microsoft/TypeScript/issues/13320

**Comment by JoshuaKGoldberg** [comment]
This has since been fixed. 🎉
